Issue355 new resets

kerimoyle opened the pull request:
Updating all parts of the code reflect the changes to the `reset` format and the removal of `when` elements (issue #355).

- [x] reset description
- [x] parser
- [x] printer
- [x] validator
- [x] coverage

NB: Reset order uniqueness (as discussed in #326) will be done in a separate PR

kerimoyle committed:
Parser tests passing

kerimoyle committed:
Printer tests passing

kerimoyle committed:
Formatting tweaks, python tests failing

kerimoyle committed:
Python tests added, python_test_docstrings failing

kerimoyle committed:
Added docstrings so python tests pass

kerimoyle commented:
**1. Reset order specification**
Currently we have code in the validator which checks for resets with non-unique order values within a component.  It was my understanding that the order should be unique within the connected variable set, not the component?  

```
if (component->resetCount() > 0) {
        // Check for duplicate order values in resets
        std::vector<int> resetOrders;
        for (size_t i = 0; i < component->resetCount(); ++i) {
            ResetPtr reset = component->reset(i);
            int resetOrder = reset->order();
            if (reset->isOrderSet()) {
                if (std::find(resetOrders.begin(), resetOrders.end(), resetOrder) != resetOrders.end()) {
                    ErrorPtr err = std::make_shared<Error>();
                    err->setDescription("Component '" + component->name() + "' contains multiple resets with order '" + convertIntToString(resetOrder) + "'.");
                    err->setComponent(component);
                    err->setRule(SpecificationRule::RESET_ORDER);
                    mValidator->addError(err);
                } else {
                    resetOrders.push_back(resetOrder);
                }
            }
        }
        for (size_t i = 0; i < component->resetCount(); ++i) {
            ResetPtr reset = component->reset(i);
            validateReset(reset, component);
        }
    }
```

**2. test_variable component?**
In the old spec document it said that the `variable` needed to be present within the parent `component`. Is this also true for the `test_variable`, or can that be in another `component`?

**3. MathML stuff**
There is a function `validator->validateMath(const std::string &input, const ComponentPtr &component)` which tests the mathML against the `component` specified.  Do the contents of the `test_value` and `reset_value` blocks need to be specific to the parent component or can it refer to variables in other components?

MichaelClerx commented:
For 2&3: Like all cellml constructs, it refers to variables in the local component only (although they may be defined in other components and then "connected" to the local versions)

What I _wanted_ to say in the earlier version, was that they had to be _defined_ (i.e. with mathml or an initial value) in the local component. But that's not allowed so we're not doing that!

kerimoyle committed:
Validator for resets

Work in progress, needing clarification on a few things before completing the parts of the validation

kerimoyle commented:
> that's not allowed

??

MichaelClerx commented:
> > that's not allowed
> 
> ??

Eh, we've had discussions about this. Basically we're saying you can write algebraic ODEs etc. in CellML, so component A can say "X + Y  = 1 / Z", and then an entirely different model can have components defining 2 out of 3 of those, and that's all valid and something tools will have to put up with.

In practice most models and tools will prefer models written as assignments only, but for the standard it's easy enough to allow [DAEs](https://en.wikipedia.org/wiki/Differential-algebraic_system_of_equations), which have some important applications

kerimoyle committed:
Coverage tests for component.cpp

kerimoyle commented:
OK, that makes sense, thanks.  Any thoughts on the scope where `order`s must be unique?  Component-wide or equivalent-variable-set wide?

MichaelClerx commented:
equivalent variable set. Are we unclear about this in the spec? We'll need to change that then...

kerimoyle commented:
The spec seemed clear to me, but the existing tests reported something completely different (as in code above).  I think it would make more sense to move the testing for order uniqueness into the equivalent variable testing routines (like where units are tested etc) instead of just looking for different resets within a component with the same order...?

kerimoyle committed:
Coverage tests for printer.cpp

MichaelClerx commented:
That sounds sensible to me. You need to have the DAG worked out before you can validate it

kerimoyle committed:
Coverage for parser.cpp

kerimoyle committed:
Coverage for reset.cpp

kerimoyle committed:
Coverage for reset.cpp (take 2)

kerimoyle commented:
DAG? Dependency Adjacency Graph? Or ...?

MichaelClerx commented:
Directed acyclic graph, sorry! The variable equivalence graph is the name we were using I think

kerimoyle committed:
Model tests fixed

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle commented:
@nickerso @agarny @hsorby Ready for review please.

nickerso committed:
Update doc strings to match code

nickerso commented on src/printer.cpp:
```diff
@@ -139,6 +137,8 @@ std::string printConnections(const ComponentMap &componentMap, const VariableMap
 std::string printMath(const std::string &math, const std::string &indent)
 {
+    // TODO This becomes unstable: need to remove the automatic addition of the indent so that mathml which
+    // has been parsed from text which already has whitespace in it doesn't get out of control
```
good to create an issue for this and list the issue here. Don't think this needs to hold up this PR.

kerimoyle replied:
Done, issue #364

nickerso commented on src/printer.cpp:
```diff
@@ -351,56 +351,50 @@ std::string Printer::PrinterImpl::printVariable(const VariablePtr &variable, con
-    if (hasValue) {
-        if (!hasCondition) {
+        // TODO Not happy with how this works - should be more consistent with other calls to printMath but that function
+        // adds extra spaces somehow ... ?
```
as above, good to link to the issue for future reference.

kerimoyle replied:
Done, issue #364

nickerso commented on src/xmlattribute.cpp:
```diff
@@ -62,6 +62,7 @@ std::string XmlAttribute::namespaceUri() const
 bool XmlAttribute::isType(const char *name, const char *ns)
 {
     bool found = false;
+
```
seems like a random change that should be taken out of this PR.

kerimoyle replied:
Done

nickerso commented on src/xmlnode.cpp:
```diff
@@ -19,6 +19,7 @@ limitations under the License.
 #include "xmlnode.h"
 
 #include <algorithm>
+#include <iostream>
```
not clear why this is here? should remove from PR.

kerimoyle replied:
Done (debugging leftovers)

nickerso commented on src/xmlnode.cpp:
```diff
@@ -68,6 +69,7 @@ bool XmlNode::isElement(const char *name, const char *ns)
         && (xmlStrcmp(reinterpret_cast<const xmlChar *>(namespaceUri().c_str()), reinterpret_cast<const xmlChar *>(ns)) == 0)
         && (xmlStrcmp(mPimpl->mXmlNodePtr->name, reinterpret_cast<const xmlChar *>(name)) == 0)) {
         found = true;
+    } else {
```
and this one.

kerimoyle replied:
Done (debugging leftovers!)

nickerso commented on tests/parser/file_parser.cpp:
```diff
@@ -25,140 +25,140 @@ limitations under the License.
-    a = model->component("c2")->math();
-    EXPECT_EQ(e2, a);
-}
+// TEST(Parser, parseSineModelFromFile)
```
Why are all these tests being commented out?

kerimoyle replied:
Back in now

nickerso commented on tests/parser/libxml_user.cpp:
```diff
@@ -24,41 +24,41 @@ limitations under the License.
-    const std::string e =
-        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
-        "<model xmlns=\"http://www.cellml.org/cellml/2.0#\"/>\n";
+// TEST(Parser, parseValidXmlDirectlyUsingLibxml)
```
more tests being commented out??

kerimoyle replied:
Back in now

nickerso commented on tests/parser/parser.cpp:
```diff
@@ -1471,23 +1470,26 @@ TEST(Parser, parseIdsOnEverything)
-        "      <when order=\"5\" id=\"w1id\">\n"
+        "    <variable name=\"variable2\" units=\"blob\" id=\"v2id\"/>\n"
+        "    <reset variable=\"variable1\" test_variable=\"variable2\" order=\"1\" id=\"r1id\">\n"
+        "      <test_value>\n"
```
Not sure if I have missed something in the discussion, but I'm expecting the test_value and rest_value to be an expression not an equation. i.e.,
```
<test_value>
   <math>
     <cn cellml:units="variable1units">3.4</cn>
   </math>
</test_value>
```

kerimoyle replied:
@nickerso I haven't seen anything either way, other than the contents of test_value and reset_value was to be mathml.  Is there a new version of the spec around resets yet? I've just been going from the use case document?  I'd imagine that the test_value needs to be an expression though? and that there could be cases where you'd want to reset to something related to another variable, so that the reset_value needs to be an expression too?  @MichaelClerx what are your thoughts?

nickerso commented on tests/printer/printer.cpp:
```diff
@@ -117,6 +117,7 @@ TEST(Printer, printEmptyReset)
 
     libcellml::Printer printer;
     const std::string a = printer.printModel(m);
+
```
not sure why this is here? another random whitespace addition to the PR that should be removed...

kerimoyle replied:
Done

nickerso commented on tests/reset/reset.cpp:
```diff
@@ -41,138 +41,103 @@ TEST(Reset, order)
-    libcellml::WhenPtr child2 = std::make_shared<libcellml::When>();
-    libcellml::WhenPtr child3 = std::make_shared<libcellml::When>();
-    libcellml::WhenPtr child4 = std::make_shared<libcellml::When>();
+// TEST(Reset, removeWhenMethods)
```
better to remove old code than to comment out.

kerimoyle replied:
Done

nickerso commented on tests/reset/reset.cpp:
```diff
@@ -201,203 +166,69 @@ TEST(Reset, printResetWithVariable)
+    libcellml::VariablePtr v1 = std::make_shared<libcellml::Variable>();
+    libcellml::VariablePtr v2 = std::make_shared<libcellml::Variable>();
+    libcellml::ResetPtr r1 = std::make_shared<libcellml::Reset>();
+    libcellml::ResetPtr r2 = std::make_shared<libcellml::Reset>();
```
unused?

kerimoyle replied:
Done

nickerso commented on tests/reset/reset.cpp:
```diff
@@ -201,203 +166,69 @@ TEST(Reset, printResetWithVariable)
+    libcellml::VariablePtr v2 = std::make_shared<libcellml::Variable>();
+    libcellml::ResetPtr r1 = std::make_shared<libcellml::Reset>();
+    libcellml::ResetPtr r2 = std::make_shared<libcellml::Reset>();
+    libcellml::ResetPtr r3 = std::make_shared<libcellml::Reset>();
```
unused?

kerimoyle replied:
Done

nickerso commented on tests/reset/reset.cpp:
```diff
@@ -476,72 +313,65 @@ TEST(Reset, addRemoveResetFromComponentMethods)
+    libcellml::VariablePtr v2 = std::make_shared<libcellml::Variable>();
     libcellml::ResetPtr r1 = std::make_shared<libcellml::Reset>();
     libcellml::ResetPtr r2 = std::make_shared<libcellml::Reset>();
     libcellml::ResetPtr r3 = std::make_shared<libcellml::Reset>();
```
unused?

kerimoyle replied:
Umm .. no?

nickerso commented on tests/reset/reset.cpp:
```diff
@@ -201,203 +166,69 @@ TEST(Reset, printResetWithVariable)
 }
 
-TEST(Reset, printResetWithMultipleWhensWithValues)
+TEST(Reset, printResetWithOrder)
```
would be good to add tests with negative orders.

kerimoyle replied:
There will be more tests around orders to come when the order checking is in properly - at the moment in this PR its scope is not correct. Related to issue #326 (and discussion earlier in this PR).

kerimoyle replied:
Order validation is now present in the other PR #350.  Tests will need to be tweaked when this is merged in, so if it's ok will add more tests when they don't need rewriting?

kerimoyle replied:
I've added a test with a negative order to the validator in this branch.

nickerso commented on tests/validator/validator.cpp:
```diff
@@ -852,143 +857,118 @@ TEST(Validator, integerStrings)
+    r7->setTestVariable(v2);
+    r7->setOrder(7);
+    r7->setResetValue(goodMath);
+    // r7->setTestValue(emptyMath);
```
would be good to test both options of either not setting the math or setting with an empty string.

kerimoyle replied:
@nickerso  I've added a test for whitespace only, which returns an equivalent error to reset-value-not-set etc ... but am confused as to whether empty math blocks should be valid or not?  See issue #365 @MichaelClerx

kerimoyle replied:
Done, including empty maths as per discussion in #365.

nickerso reviewed (dismissed):
Looking good. A bit more work to clarify things, test some more combinations, and put the parsing tests back in...

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
Lots of tidying up, tests uncommented, math printing still WIP

- math parse and printing fixed
- tests uncommented
- removed validation from parser of resets
- added validation for empty string maths block
- maths parse->print->parse not working for invalid maths blocks: extra whitespace is added

kerimoyle committed:
formatting tweaks

kerimoyle commented:
> Directed acyclic graph, sorry! The variable equivalence graph is the name we were using I think

@MichaelClerx Is it directed?  Ooops, I'd thought not ... ? Might need to adjust validation checks if so!

kerimoyle committed:
Add validation that variable and test_variable are inside the parent component

kerimoyle committed:
formatting

kerimoyle commented:
Added validation for reset variable and test_variable being inside the same component, and the tests to cover them. Addresses issue #294.

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
Empty maths blocks are valid

kerimoyle committed:
formatting

kerimoyle committed:
Added test with negative reset order

hsorby reviewed (changes requested):
TEST(Validator, resets) is way too big and it needs to be broken down into smaller tests.

kerimoyle commented:
> TEST(Validator, resets) is way too big and it needs to be broken down into smaller tests.

OK, will do.  Micro-tests coming up ... ;)

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
Split reset validation tests into smaller bits

kerimoyle committed:
Coverage for validator

kerimoyle commented:
@hsorby @agarny @nickerso I think I've addressed everything requested ... Good to go?

hsorby reviewed (changes requested):
We also need to have some printer tests for making sure we print out resets as we expect.

hsorby commented:
The broken down tests are much better for me I find them much easier to read and make sense of.

kerimoyle commented:
Not at all DRY though :\

hsorby commented:
Nope.  But it's testing so essentially you want to repeat yourself.

kerimoyle commented:
> We also need to have some printer tests for making sure we print out resets as we expect.

We already test printing in the tests/reset.cpp for:

- printResetWithVariable
- printResetWithoutTestValue
- printResetWithOrder
- printResetWithNegativeOrder
- printResetWithOrderAndVariable
- addRemoveResetFromComponentMethods

... and in tests/printer.cpp for:
- printEmptyReset

I just adjusted the existing ones to the new format from what was there previously.  Are there tests that need to be added to these, or should the reset ones be in the printer.cpp file?

hsorby commented:
There is no right place for them I was just trying to find them and thought they would be in the `tests/printer/printer.cpp` file but I didn't see any changes there.  I'll have another look.

hsorby commented on tests/parser/parser.cpp:
```diff
@@ -1531,23 +1532,18 @@ TEST(Parser, parseResets)
-        "      <when order=\"5\">\n"
+        "    <variable name=\"variable2\" id=\"vid2\"/>\n"
+        "    <reset order=\"1\" id=\"rid\" variable=\"variable1\" test_variable=\"variable2\">\n"
+        "      <test_value>\n"
```
Everything in this test should have an id, this doesn't.

kerimoyle replied:
Is that a general thing for cellml stuff everywhere?  I thought ids were optional?

kerimoyle replied:
Ignore me, just read the name of the test it's in ... :)

kerimoyle replied:
Done

hsorby commented on tests/parser/parser.cpp:
```diff
@@ -1531,23 +1532,18 @@ TEST(Parser, parseResets)
         "          some condition in mathml\n"
         "        </math>\n"
+        "      </test_value>\n"
+        "      <reset_value>\n"
```
This should also have an id.

kerimoyle replied:
Done

hsorby commented on tests/parser/parser.cpp:
```diff
@@ -1560,152 +1556,175 @@ TEST(Parser, parseResets)
+TEST(Parser, parseResetsWithErrors)
 {
-    const std::string input =
+    /* const std::string input =
```
This shouldn't be here if it's just commented out.

hsorby commented on tests/parser/parser.cpp:
```diff
@@ -1560,152 +1556,175 @@ TEST(Parser, parseResets)
 
-    EXPECT_EQ(resetExpected, parser.error(0)->reset());
-    EXPECT_EQ(whenExpected, parser.error(1)->when());
+    std::cout << printer.printModel(model);
```
This has to go.

kerimoyle replied:
Yep

hsorby commented on tests/parser/parser.cpp:
```diff
@@ -1785,3 +1806,131 @@ TEST(Parser, xmlComments)
+    EXPECT_EQ(input, output3);
+}
+
+// TODO Removed so that math print/parse behaviour can be investigated independently through #364
```
Not a great job of removing???

kerimoyle replied:
True dat ... I left it commented out because of the weirdness I was getting in parsing maths between two branches.

kerimoyle replied:
Deleted now that I know that the mathml printing/parsing is inconsistent for invalid maths.

hsorby commented:
Trying to find resets API tests `appendResetValue` and so on but haven't managed it can someone point them out to me?

hsorby commented:
Also do we want to have a `clearResetValue` API etc?

kerimoyle commented:
> Also do we want to have a `clearResetValue` API etc?

I took the current implementation from what was there previously (see component's mathml functions?).   The same behaviour is found by calling `setResetValue("");` . Easy enough to add, but then will also be needed throughout as well?

kerimoyle commented:
> Trying to find resets API tests `appendResetValue` and so on but haven't managed it can someone point them out to me?

TEST(Parser, repeatedMathParsePrintBehaviourWithReset)  calls code that uses them.  Does it need to be separately done?

hsorby commented:
> > Also do we want to have a `clearResetValue` API etc?
> 
> I took the current implementation from what was there previously (see component's mathml functions?). The same behaviour is found by calling `setResetValue("");` . Easy enough to add, but then will also be needed throughout as well?

No I'm pretty sure we decided that setting empty is clear.  That is what everything else does.  Just reminding myself of that.

hsorby commented:
> TEST(Parser, repeatedMathParsePrintBehaviourWithReset) calls code that uses them. Does it need to be separately done?

I would have that done they are unit tests each part should be tested on it's own.

kerimoyle committed:
Added ids to reset_value and test_value in resets

- parsed
- printed
- tests updated

kerimoyle committed:
docstrings

kerimoyle committed:
More tests for resets

kerimoyle commented:
> I would have that done they are unit tests each part should be tested on it's own.

@hsorby Have added four new tests into tests/reset.cpp which deal with each bit individually: 
- testValueSetClear
- testValueAppend
- resetValueSetClear
- resetValueAppend

What's next ;) ?

hsorby commented on tests/reset/reset.cpp:
```diff
@@ -360,3 +360,266 @@ TEST(Reset, hasResetFromComponentMethod)
+
+TEST(Reset, resetValueSetClear)
+{
+    static const std::string emptyMath =
```
This should be moved into the test_utils.h file and then used throughout the tests.

kerimoyle replied:
Good plan.

kerimoyle replied:
Done

hsorby commented on tests/reset/reset.cpp:
```diff
@@ -360,3 +360,266 @@ TEST(Reset, hasResetFromComponentMethod)
+        "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
+        "</math>\n";
+
+    const std::string in1 =
```
Not used so delete

kerimoyle replied:
Done

hsorby commented on tests/reset/reset.cpp:
```diff
@@ -360,3 +360,266 @@ TEST(Reset, hasResetFromComponentMethod)
+    static const std::string firstMaths = "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n";
+    static const std::string secondMaths = "</math>\n";
+
+    const std::string in =
```
Not used so delete.

kerimoyle replied:
Done

hsorby commented on tests/reset/reset.cpp:
```diff
@@ -360,3 +360,266 @@ TEST(Reset, hasResetFromComponentMethod)
+
+TEST(Reset, resetValueAppend)
+{
+    static const std::string firstMaths = "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n";
```
Could also choose to append two empty maths blocks.

kerimoyle replied:
Oh, I could have chosen many, many other things ... :)

hsorby commented on tests/reset/reset.cpp:
```diff
@@ -360,3 +360,266 @@ TEST(Reset, hasResetFromComponentMethod)
+        "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
+        "</math>\n";
+
+    const std::string in1 =
```
Not used so delete

hsorby reviewed (changes requested):
A few things to tidy up.

kerimoyle commented:
I'm missing loads of variables which are declared but unused because there used to be a warning for this which I relied on ... why was it removed?  Seems a more useful warning than the your-test-failed-because-you-missed-a-single-space-tests. Or does anyone know how I do that locally?

kerimoyle committed:
Removed unused variables, emptyMath to test_utils.h

kerimoyle commented:
Good to go? @hsorby @nickerso @agarny

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
clang format tidied

kerimoyle committed:
Merge pull request #18 from cellml/develop

updating from upstream

agarny commented on src/api/libcellml/reset.h:
```diff
@@ -57,107 +57,112 @@ class LIBCELLML_EXPORT Reset: public OrderedEntity
+     * @brief Set the @c TestVariable for this @c Reset.
      *
-     * Add a copy of the given @p when as a child @c When of this @c Reset.
+     * Sets the @c TestVariable for this @c Reset.
```
Use the infinitive, i.e. "Set" rather than "Sets". (Check the rest of the PR for similar occurrences.)

kerimoyle replied:
Done :)

kerimoyle committed:
Conjugation of comments

agarny commented on src/api/libcellml/reset.h:
```diff
@@ -57,107 +57,112 @@ class LIBCELLML_EXPORT Reset: public OrderedEntity
-     * @brief Remove the @c When with the given pointer.
-     *
-     * Remove the @c When with the pointer @p when.
+     * @brief Appends the argument to the testValue string for this reset.
```
Replace with something like "@brief Append @p math to the existing testValue string for this @c Reset." (i.e. try to be consistent with what has been done for the `Component` class.) The same for the other API documentation.

kerimoyle replied:
I think we should do this en masse for all of the documentation everywhere once all the pull requests have been merged.  There are lots of inconsistencies everywhere ... !

agarny replied:
Agreed, ignore my comment for now.

kerimoyle commented:
Nudge ... @hsorby @nickerso

hsorby committed:
Modify the documentation around resets.  Minor tidying of reset test code.

hsorby reviewed (commented):
Looks good to me I have made some documentation changes and made a pull request [here](https://github.com/kerimoyle/libcellml-keri/pull/19) to (maybe) address issues I have with the code.

hsorby commented:
Sorry it took so long to get here.

kerimoyle committed:
Merge pull request #19 from hsorby/new_resets_validation_i355

Reviewed code

kerimoyle commented:
I've merged in @hsorby 's  changes now - all around comments.  Once all the outstanding PRs are into development it would be good to go through the comments again (see @agarny 's earlier comment here ...)

kerimoyle commented:
So this one is good to merge now :) @hsorby @nickerso @agarny 
ps - not sure why the earlier reviews are still showing as requested, all items are finished.

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
Merge branch 'issue355_new_reset_validation' of https://github.com/kerimoyle/libcellml-keri into kerimoyle-issue355_new_reset_validation

Merging changes from other recent PRs into this one

kerimoyle committed:
Merge branch 'kerimoyle-issue355_new_reset_validation' into develop

kerimoyle commented:
Help ... not sure what's going on here ... I've merged the updated development branch into this one just fine by following the instructions here, but now I get a unit test failure on the mergability test?  All tests are passing on my machine and my local branches merged without a hitch.  What does that mean? @hsorby @agarny

agarny commented:
You might want to try to merge one more time. From experience, GitHug can sometimes report conflicts when there are actually none. It happened to me earlier today with [@hsorby's PR](https://github.com/agarny/libcellml/pull/5).

kerimoyle committed:
Update parser.cpp

Fake changes to trigger commit

kerimoyle committed:
Revert "Update parser.cpp"

This reverts commit cab41dc7b6bd113afbd0c1e413527d09b1614c1d.

kerimoyle committed:
Revert "Merge branch 'issue355_new_reset_validation' of https://github.com/kerimoyle/libcellml-keri into kerimoyle-issue355_new_reset_validation"

This reverts commit 78ed057150910a6211b7ff1e4ad7f276695aa58c, reversing
changes made to 411ff0b8a57e53e0cbf340701f92e219f8f9f480.

kerimoyle committed:
Merge remote-tracking branch 'upstream/develop' into develop

kerimoyle committed:
Update validator.cpp

kerimoyle committed:
Update validator.cpp

kerimoyle committed:
Mass update from working branch

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle commented:
Whew.  Should be good to merge now? @agarny @hsorby @nickerso

kerimoyle commented:
Nudge nudge ... @hsorby @agarny @nickerso

hsorby commented:
Looking at it now. :)

hsorby commented:
There are some TODO's in the parser that mention removing becasuse they are validation.  I think they should definitely be removed and other cases also considered for removal.

kerimoyle commented:
Thanks @hsorby I'll have another look.  There was a PR at some stage which removed these but I can't find it anymore ... :|

kerimoyle commented:
... and which other cases are you talking about?

hsorby commented:
Well ... there are some errors produced from invalid numbers for instance.  I can't decide if this is a more of a validation problem or something we should report that we have changed that may be a surprise to the user thus worthy of informing.

kerimoyle commented:
I'm just going through the current `parser.cpp `file now and will remove anything that's not related to parsing.  In designing the Issue class I've split up the ideas based on whether people would get an informative error message from the validator alone - there were places in the parser where it made more sense to flag errors as they don't get propagated in a helpful way later.  I'll comment out all the tests and lines which are currently throwing errors here just to check that we're on the same page, then I'll remove them properly.  But .. this should all be part of the bigger discussion about what kind of things and where we want the `Issue` class to report ...?

kerimoyle committed:
Removing validation from parser

... commented out for now to see what others think ...

kerimoyle committed:
Update tests in connection.cpp

kerimoyle committed:
Update coverage.cpp

Commented out the effects of taking validation out of the parser

kerimoyle commented:
So I've gone back and checked the other messages, in particular around the number formatting stuff.  Even though it's technically validation to check the number format (see below), since the numbers are only _saved_ if they're valid, the validator has no way to access whatever the wrong value was since it wouldn't be saved.  This is one of those examples where once we have the `Issue` class it will make more sense - we can raise a warning in the parser instead of ignoring it altogether ... 

```
        if (attribute->isType("units")) {
            reference = attribute->value();
        } else if (attribute->isType("prefix")) {
            prefix = attribute->value();
        } else if (attribute->isType("exponent")) {
            if (isCellMLReal(attribute->value())) {
                exponent = convertToDouble(attribute->value());
            } else {
                ErrorPtr err = std::make_shared<Error>();
                err->setDescription("Unit referencing '" + node->attribute("units") + "' in units '" + units->name() + "' has an exponent with the value '" + attribute->value() + "' that is not a representation of a CellML real valued number.");
                err->setUnits(units);
                err->setRule(SpecificationRule::UNIT_EXPONENT);
                mParser->addError(err);
            }
        } else if (attribute->isType("multiplier")) {
            if (isCellMLReal(attribute->value())) {
                multiplier = convertToDouble(attribute->value());
            } else {
                ErrorPtr err = std::make_shared<Error>();
                err->setDescription("Unit referencing '" + node->attribute("units") + "' in units '" + units->name() + "' has a multiplier with the value '" + attribute->value() + "' that is not a representation of a CellML real valued number.");
                err->setUnits(units);
                err->setRule(SpecificationRule::UNIT_MULTIPLIER);
                mParser->addError(err);
            }
```

kerimoyle committed:
Update parser.cpp

Commented out all validation code for checking

kerimoyle committed:
Tweaking parser errors

kerimoyle commented:
So I've come full circle!  There are a lot of things throughout the parser which we need to figure out how to handle, and this can only be done in a helpful way once the `Issue` class is done.  I think they should stay as-is for the minute and be addressed separately so that it's not confused with the reset format which is the point of this PR?

kerimoyle committed:
Update parser.cpp

kerimoyle committed:
Parser errors reinstated

Removed the check for missing order in resets, all others follow the same pattern as errors in encapsulation, or are not able to be reported well in the validator.

kerimoyle commented:
I think it's ready now.  Here's the long story:
- I've put back the previously commented-out error creation in the parser which deal with encapsulations as this PR should only deal with the resets.  The others will be dealt with in #314 and #368 
- I've removed the check for the order in a parsed reset being set, as this is purely validation and can be picked up later
- I've reinstated the checks for validation errors when the **real** cause of that error cannot be tracked downstream, ie: when an invalid piece of text is not saved, so the error that the validator can report is a lot less informative than that which the parser can report.  Things in this category are the number errors (the validator can only say that they're missing, not what's wrong with them, as their value is not saved).
- the other reset errors in the parser follow the same pattern as the parser errors elsewhere, most closely the parsing of encapsulations.  If we want to remove them from parsing the resets, we should consider removing them from parsing the encapsulations ... but that's a story for another PR like the one above.

Short story ... I think it's good to go now?  @hsorby @agarny @nickerso

kerimoyle commented:
Nudge, nudge ... @hsorby @agarny @nickerso I think this one should go in before the #360 code generation one so that the generated code can be the right format?

kerimoyle committed:
Update validator.cpp

kerimoyle committed:
Merge remote-tracking branch 'upstream/develop' into develop

kerimoyle committed:
Updating to origin

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
Merge pull request #25 from cellml/develop

update from upstream

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
Revert "Merge branch 'develop' into issue355_new_reset_validation"

This reverts commit 38d155131f63224c22316fb402a03f21281dfed7, reversing
changes made to 2ead5e32120c553b69186b56c0025b5046713297.

kerimoyle commented:
Pretty pretty please can someone review this ... @hsorby @agarny @nickerso

kerimoyle committed:
Merge remote-tracking branch 'upstream/develop' into develop

kerimoyle committed:
Merge branch 'develop' into issue355_new_reset_validation

kerimoyle committed:
ignore me - trying to get build

kerimoyle committed:
Merge conflicts
